types/accessRequest.js

```diff
@@ -0,0 +1,76 @@
+const Serializable = require('./serializable')
+
+class AccessRequest extends Serializable {
+  constructor(reason, requestorID, realmName, groups, accessDurationSeconds) {
+    super()
+    this.reason = reason
+    this.requestorID = requestorID
+    this.realmName = realmName
+    this.groups = groups
+    this.accessDurationSeconds = accessDurationSeconds
+    this.ID = null
+    this.state = null
+    this.createdAt = null
+    this.autoExpiresAt = null
+  }
+
+  serializable() {
+    let toSerialize = {
+      reason: this.reason,
+      requestor_id: this.requestorID,
+      realm_name: this.realmName,
+      ttl: this.accessDurationSeconds,
+      groups: [],
+      id: this.ID,
+      state: this.state,
+      created_at: this.createdAt,
+      auto_expires_at: this.autoExpiresAt,
+    }
+    let rawGroups = []
+    for (const group of this.groups) {
+      rawGroups.push({ group_id: group.ID })
+    }
+    toSerialize.groups = rawGroups
+    const serializedKeys = Object.keys(toSerialize)
+    for (const key of serializedKeys) {
+      if (toSerialize[key] === null) {
+        delete toSerialize[key]
+      }
+    }
+    return toSerialize
+  }
+  static decode(json) {
+    let reason = json.reason === undefined ? null : json.reason
+    let requestorID = json.requestor_id === undefined ? null : json.requestor_id
+    let realmName = json.realm_name === undefined ? null : json.realm_name
+    let rawGroups = json.groups === undefined ? null : json.groups
+    let groups = []
+    for (const group of rawGroups) {
+      groups.push({ ID: group.group_id })
+    }
+    let accessDurationSeconds = json.ttl === undefined ? null : json.ttl
+
+    var accessRequest = new AccessRequest(
+      reason,
+      requestorID,
+      realmName,
+      groups,
+      accessDurationSeconds
+    )
+
+    // server defined values
+    let ID = json.id === null ? null : json.id
```

Comment:
could we simplify these null checks: `let ID = json.id || null`

I think the existing code can return undefined due to the explicit null check, maybe we want that instead of always setting null?

Reply:
i'd just use `json.id || null` for now. you just need to be aware that `0`, `""`, & `false` are falsy so would both (sometimes incorrectly) fall back to `null` but i don't think that should matter for these values. 

the only other thing i'd change here is to assign directly to the object rather than having these intermediary variables